packages/test/test-end-to-end-tests/src/test/dataMigration.spec.ts

```diff
@@ -0,0 +1,425 @@
+	await runningSummarizer.summarizingLock;
+	await runningSummarizer.lockedSummaryAction(
+		// eslint-disable-next-line @typescript-eslint/no-unsafe-return
+		() => runningSummarizer.beforeSummaryAction(),
+		async () => {
+			const result = await summarizerRuntime.submitSummary({
+				fullTree: true,
+				summaryLogger: new TelemetryNullLogger(),
+				cancellationToken: {
+					cancelled: false,
+					waitCancelled: neverCancel.promise,
+				},
+				summarizeResult: appSummary,
+			});
+			assert(result.stage === "submit", "should have submitted summary");
+		},
+		// eslint-disable-next-line @typescript-eslint/no-unsafe-return
+		() => runningSummarizer.afterSummaryAction(),
+	);
+	summarizerRuntime.on("op", (op) => {
+		if (op.type === MessageType.Accept) {
+			summarizerRuntime.disposeFn();
+		}
+	});
+	const waitForAck = new Deferred<ISummaryAck>();
+	summarizerRuntime.on("op", (op) => {
+		if (op.type === MessageType.SummaryAck) {
+			waitForAck.resolve(op.contents);
+		}
+	});
+	const ack = await waitForAck.promise;
+	summarizerRuntime.submitFinished("summaryHandle");
+	return ack;
+}
+
+describeNoCompat("Data Migration is possible", (getTestObjectProvider) => {
+	let provider: ITestObjectProvider;
+	const runtimeOptions: IContainerRuntimeOptions = {
+		summaryOptions: {
+			summaryConfigOverrides: {
+				...DefaultSummaryConfiguration,
+				...{
+					maxOps: 1,
+					initialSummarizerDelayMs: 0,
+					minIdleTime: 10,
+					maxIdleTime: 10,
+					nonRuntimeHeuristicThreshold: 1,
+				},
+			},
+		},
+	};
+	const dataObjectFactoryV1 = new DataObjectFactory(
+		TestDataObjectType,
+		DataObjectV1,
+		[SharedCell.getFactory()],
+		[],
+	);
+
+	const dataObjectFactoryV1ToV2 = new DataObjectFactory(
+		TestDataObjectType,
+		DataObjectV2,
+		[new SharedCellToSharedMapFactory()],
+		[],
+	);
+	const dataObjectFactoryV2 = new DataObjectFactory(
+		TestDataObjectType,
+		DataObjectV2,
+		[SharedMap.getFactory()],
+		[],
+	);
+	const runtimeFactoryV1 = new ContainerRuntimeFactoryWithDefaultDataStore(
+		dataObjectFactoryV1,
+		[[dataObjectFactoryV1.type, Promise.resolve(dataObjectFactoryV1)]],
+		undefined,
+		undefined,
+		runtimeOptions,
+	);
+
+	const runtimeFactoryV1ToV2 = new ContainerRuntimeFactoryWithDefaultDataStore(
+		dataObjectFactoryV1ToV2,
+		[[dataObjectFactoryV1ToV2.type, Promise.resolve(dataObjectFactoryV1ToV2)]],
+		undefined,
+		undefined,
+		runtimeOptions,
+	);
+
+	const runtimeFactoryV2 = new ContainerRuntimeFactoryWithDefaultDataStore(
+		dataObjectFactoryV2,
+		[[dataObjectFactoryV2.type, Promise.resolve(dataObjectFactoryV2)]],
+		undefined,
+		undefined,
+		runtimeOptions,
+	);
+
+	const deferred = new Deferred<{
+		summarizer: ContainerRuntime;
+		readonlyRuntime: ContainerRuntime;
+	}>();
+
+	const dataMigrationRuntimeFactoryV2 = new ContainerRuntimeFactoryWithDataMigration(
+		runtimeFactoryV1,
+		runtimeFactoryV1ToV2,
+		runtimeFactoryV2,
+		deferred,
```

Comment:
I'd suggest instead have a function on ContainerRuntimeFactoryWithDataMigration that returns this promise. Under cover (inside class) we can use same deferred approach, but having a function with a name and a call many lines below will be more readable as one could glimpse at least some context from the name - when I see where this differed is used 40 lines down, I have no context where it came from, and what are the semantics for it - need to drill into implementation of this class to get that.